Load space objects into mission

Dark-Visor opened the pull request:
Needs some reviewing and guidance. Thanks

Dark-Visor committed:
Missionparse work for objects

Very much a work in progress. I have tried to comment as much as
possible to make my thinking clearer.

When the asteroids are tabled in FotG we can point the cases to them.
The closest mine is currently the ION mine but the weapon can be changed
in FRED.

As for the backgrounds. It would probably be best to not process these
at all from the game and perhaps at a later stage, add FotG backgrounds
depending on where the location of the mission is - from the briefing.

For the mine field, because it is in a 2d plane with each mine having
coordinates (a,b) I felt the best way was to have two for loops... the
first would move one step along one axis, then the inner for loop would
populate the mines along the perpendicular axis. The grid is populated
starting on the initial mine as the origin however in XWing the mine is
in the centre of the grid. Therefore the second part of the calculation
is to shift the mine back so centre the grid on the given position.

This second part of the calculation could be done before the for loops
which means it would only have to be done once and would also help make
the code clearer.

Dark-Visor committed:
MissionParse Object update to make 2d minefiled

I changed the code for the calculation of the minefiled just to make it
a little bit clearer as well as removing some redundancy.

The centre of the grid is now calucluated before the minefield is
populated rather than before when the whole minefield was shifted
afterwards.

Goober5000 commented on code/mission/import/xwingmissionparse.cpp:
```diff
@@ -623,6 +623,234 @@ void parse_xwi_flightgroup(mission *pm, const XWingMission *xwim, const XWMFligh
 	}
 }
 
+bool space_object = true; // true for objects, false for backgrounds - needs to be checked
```
There shouldn't be a global variable here.  The information should be able to be determined some other way

Dark-Visor replied:
Resolved with the fix here https://github.com/Fate-of-the-Galaxy/fs2open.github.com/pull/3#discussion_r1370751386 ?

Goober5000 replied:
Yeah, that is the way that objects are determined.

Goober5000 commented on code/mission/import/xwingmissionparse.cpp:
```diff
@@ -623,6 +623,234 @@ void parse_xwi_flightgroup(mission *pm, const XWingMission *xwim, const XWMFligh
+		return " ";
+	case XWMObjectType::oj_Asteroid8:
+		return " ";
+	case XWMObjectType::oj_Rock_World:    // should I just remove these cases for the backgrounds?
```
Yeah, you can remove this case and all of the following cases in this switch statement.  The switch will revert to the default case, which just calls break, and then the function will return nullptr, indicating there's no object to be created.

Goober5000 commented on code/mission/import/xwingmissionparse.cpp:
```diff
@@ -623,6 +623,234 @@ void parse_xwi_flightgroup(mission *pm, const XWingMission *xwim, const XWMFligh
+}
+
+vec3d xwi_determine_mine_formation_position(const XWMObject* oj, float objectPosX, float objectPosY, float objectPosZ, 
+	float objectPosA, float objectPosB)
```
Call these something different from `objectPosA` and `objectPosB`; they are confusing at first glance.  Maybe `offsetAxisA` and `offsetAxisB`

Goober5000 commented on code/mission/import/xwingmissionparse.cpp:
```diff
@@ -623,6 +623,234 @@ void parse_xwi_flightgroup(mission *pm, const XWingMission *xwim, const XWMFligh
+	case XWMObjectFormation::ojf_FloorXY:
+		return vm_vec_new((objectPosX + objectPosA), objectPosZ, (objectPosY + objectPosB));
+	case XWMObjectFormation::ojf_SideYZ:
+		return vm_vec_new(objectPosX, (objectPosZ + objectPosB), (objectPosY + objectPosA));
```
Is it correct for A and B to be switched here?

Dark-Visor replied:
It shouldn't matter which one is A and B but I'll switch them so A comes first as it is neater. The reason why they are back to front is just because I switched Y and Z for FSO but didn't rename A and B for ojf_SideYZ.

Goober5000 commented on code/mission/import/xwingmissionparse.cpp:
```diff
@@ -623,6 +623,234 @@ void parse_xwi_flightgroup(mission *pm, const XWingMission *xwim, const XWMFligh
+	return vm_vec_new(objectPosX, objectPosZ, objectPosY);
+}
+
+void xwi_determine_object_pbh(matrix* orient, const XWMObject* oj)
```
Call this _orient instead of _pbh

Goober5000 commented on code/mission/import/xwingmissionparse.cpp:
```diff
@@ -623,6 +623,234 @@ void parse_xwi_flightgroup(mission *pm, const XWingMission *xwim, const XWMFligh
+		centered over their starting point.For example,
+		a value of 3 will create a 3x3 grid of mines.**/
+
+		SCP_string object_type = xwi_determine_object_type(oj);
```
This should be right under SCP_UNUSED.  It should use `auto`, not `SCP_string`, because we want to keep the type as `const char *`.  Additionally there should be an if() statement to return from the function if the type is `nullptr`.

Goober5000 commented on code/mission/import/xwingmissionparse.cpp:
```diff
@@ -623,6 +623,234 @@ void parse_xwi_flightgroup(mission *pm, const XWingMission *xwim, const XWMFligh
+
+		SCP_UNUSED(pm);
+
+		int number_of_objects = oj->numberOfObjects;
```
Add a check to return from the function if the number is < 1.

Goober5000 commented on code/mission/import/xwingmissionparse.cpp:
```diff
@@ -623,6 +623,234 @@ void parse_xwi_flightgroup(mission *pm, const XWingMission *xwim, const XWMFligh
+		float objectPosY = oj->object_y;
+		float objectPosZ = oj->object_z;
+
+		int mine_dist = 100; // change this to change the distance between the mines
```
This should be 400

Goober5000 commented on code/mission/import/xwingmissionparse.cpp:
```diff
@@ -623,6 +623,234 @@ void parse_xwi_flightgroup(mission *pm, const XWingMission *xwim, const XWMFligh
+
+				suffix = object_type.c_str() + (i * m); // ok so not sure how to name them... add suffix?
+				strcpy_s(pobj.name, suffix.c_str());    // add a suffix to the name...??
+				SCP_totitle(pobj.name);
```
For generating the name, create a routine using this naming code from ship.cpp:
```
		// regular name, regular suffix
		char base_name[NAME_LENGTH];
		char suffix[NAME_LENGTH];
		strcpy_s(base_name, Ship_info[ship_type].name);
		sprintf(suffix, NOX(" %d"), n);

		// default names shouldn't have a hashed suffix
		end_string_at_first_hash_symbol(base_name);

		// start building name
		strcpy_s(shipp->ship_name, base_name);

		// if generated name will be longer than allowable name, truncate the class section of the name by the overflow
		int char_overflow = static_cast<int>(strlen(base_name) + strlen(suffix)) - (NAME_LENGTH - 1);
		if (char_overflow > 0) {
			shipp->ship_name[strlen(base_name) - static_cast<size_t>(char_overflow)] = '\0';
		}

		// complete building the name by adding suffix number
		strcat_s(shipp->ship_name, suffix);
```
You'll need to use `pobj.name` instead of `shipp->ship_name`, and you'll also need to pick the first `n` that doesn't have a name collision.  Do this by repeatedly looping through `n` until you find one that can be used to create a name that doesn't already exist in `Parse_objects`.

Dark-Visor committed:
Update enum indices and unique naming of objects

Add values to enums in XWMObjectType in xwinglib.h to rectify gaps
between various object types.

In xwingmissionparse.cpp add code to give unique names to each object
and each object in a group using suffixes. Other cleanups.

Goober5000 commented on code/mission/import/xwingmissionparse.cpp:
```diff
@@ -109,6 +109,9 @@ int xwi_determine_arrival_cue(const XWingMission *xwim, const XWMFlightGroup *fg
 		return Locked_sexp_true;
+	
+	if (fg->arrivalEvent == XWMArrivalEvent::ae_mission_start)
+		return Locked_sexp_true;
```
It's not necessary to add this code here because it was already included in the other PR.  Since the other PR was merged, this code now exists in master.  Once you merge your branch and update master, you'll see it too.

Goober5000 commented on code/mission/import/xwingmissionparse.cpp:
```diff
@@ -623,6 +626,237 @@ void parse_xwi_flightgroup(mission *pm, const XWingMission *xwim, const XWMFligh
+	case XWMObjectType::oj_Training_Platform11:
+		return nullptr;
+	case XWMObjectType::oj_Training_Platform12:
+		return nullptr;
```
Just delete all these lines.  The function will return nullptr for any enum that isn't part of the switch, and by removing these lines you can reduce clutter and focus on the significant parts.

Goober5000 commented on code/mission/import/xwingmissionparse.cpp:
```diff
@@ -623,6 +626,237 @@ void parse_xwi_flightgroup(mission *pm, const XWingMission *xwim, const XWMFligh
+
+	// Copy objects in Parse_objects to set for name checking below
+	// This only needs to be done fully once per object group then can be added to after each new object
+	SCP_set<char[32]> objectSet;
```
Ironically, if you're using SCP_set, then you'll need to make the template type SCP_string.  The reason is that SCP_string implements the equals operator whereas char[] does not.

Also call it objectNameSet or objectNames since it's just storing names, not entire X-Wing objects or parse objects.

Goober5000 commented on code/mission/import/xwingmissionparse.cpp:
```diff
@@ -623,6 +626,237 @@ void parse_xwi_flightgroup(mission *pm, const XWingMission *xwim, const XWMFligh
+		
+	matrix orient;
+	xwi_determine_object_orient(&orient, oj);
+	p_object pobj;
```
Don't declare pobj here.  Declare it right before `pobj.orient = orient`.

Goober5000 commented on code/mission/import/xwingmissionparse.cpp:
```diff
@@ -623,6 +626,237 @@ void parse_xwi_flightgroup(mission *pm, const XWingMission *xwim, const XWMFligh
+		return;
+
+	/**
+	NumberOfCraft Holds various parameter values depending on object types.For mines,
```
I think NumberOfCraft is always 1 for space objects that aren't mines.  Add a check for this.  If the space object isn't a mine and has a number more than 1, display a `Warning` and set the number to 1.

Goober5000 commented on code/mission/import/xwingmissionparse.cpp:
```diff
@@ -623,6 +626,237 @@ void parse_xwi_flightgroup(mission *pm, const XWingMission *xwim, const XWMFligh
+				sprintf(suffix, NOX(" %d"), n);
+				strcpy_s(croppedName, base_name);
+				strcat_s(croppedName, suffix);
+			}
```
Extract all of this into a separate function.

What is `newArray`?  It's not being used, and shouldn't be needed, so take it out.

Instead of hard-coding the number 4, use `strlen(suffix)` as in the sample code.

You don't need a separate `croppedName` variable.  Use `pobj.name` itself.

Change `pobj.name` to title case as soon as it's created.  You want the name that is checked to be the same as the name that is stored in the set.

Goober5000 commented on code/mission/import/xwingmissionparse.cpp:
```diff
@@ -623,6 +626,237 @@ void parse_xwi_flightgroup(mission *pm, const XWingMission *xwim, const XWMFligh
+			if (pobj.arrival_anchor < 0)
+				pobj.arrival_location = ARRIVE_AT_LOCATION;
+			if (pobj.departure_anchor < 0)
+				pobj.departure_location = DEPART_AT_LOCATION;
```
Some of this is needed, but it should use default values.  The arrival cue should be `Locked_sexp_true`, the departure cue should be `Locked_sexp_false`, and the locations should be at-location.  The delays and anchors aren't needed.

Goober5000 commented on code/mission/import/xwingmissionparse.cpp:
```diff
@@ -623,6 +626,237 @@ void parse_xwi_flightgroup(mission *pm, const XWingMission *xwim, const XWMFligh
+		pobj.max_shield_recharge = sip->max_shield_recharge;
+		pobj.replacement_textures =
+			sip->replacement_textures; // initialize our set with the ship class set, which may be empty
+		pobj.score = sip->score;
```
Uncomment these lines.  Instead of `ai_index`, use `sip->ai_class`.

Goober5000 commented on code/mission/import/xwingmissionparse.cpp:
```diff
@@ -623,6 +626,237 @@ void parse_xwi_flightgroup(mission *pm, const XWingMission *xwim, const XWMFligh
+	centered over their starting point.For example,
+	a value of 3 will create a 3x3 grid of mines.**/
+
+		
```
Figure out the ship class somewhere near the top of the function...
```
	int ship_class = ship_info_lookup(object_type);
	if (ship_class < 0)
	{
		Warning(LOCATION, "Unable to determine ship class for Object Group with type %s", object_type);
		ship_class = 0;
	}
	auto sip = &Ship_info[ship_class];
```

Dark-Visor committed:
Separate the function to check for unique suffix

For the unique name check for the object, put this into a separate
function. Update missionparse to add default values for the object
initialisation.

Add a goto for now in case the search for a unique name gets stuck in an
infinte loop. This can potentially be removed in a later PR.

Goober5000 commented on code/mission/import/xwingmissionparse.cpp:
```diff
@@ -623,6 +623,231 @@ void parse_xwi_flightgroup(mission *pm, const XWingMission *xwim, const XWMFligh
+	case XWMObjectType::oj_Mine2:
+		return "Defense Mine#Ion";
+	case XWMObjectType::oj_Mine3:
+		return "Defense Mine#Ion";
```
You can remove the duplicate "Defense Mine#Ion" strings and just stack the 1-4 cases

Goober5000 commented on code/mission/import/xwingmissionparse.cpp:
```diff
@@ -623,6 +623,231 @@ void parse_xwi_flightgroup(mission *pm, const XWingMission *xwim, const XWMFligh
+}
+
+// find lowest unique n for the suffix should return true when a unique name is found else false
+bool unique_object_suffix(SCP_set<SCP_string> objectNameSet, char base_name[NAME_LENGTH], char suffix[NAME_LENGTH])
```
This function should contain the entire code for finding and creating a name.  This includes the loop.  The only two parameters should be the set and a `char[NAME_LENGTH]` parameter to return the name.  The `base_name` and `suffix` variables should be local to the function.

Also, the `objectNameSet` parameter should have a reference operator (`&`).

Goober5000 commented on code/mission/import/xwingmissionparse.cpp:
```diff
@@ -623,6 +623,231 @@ void parse_xwi_flightgroup(mission *pm, const XWingMission *xwim, const XWMFligh
+				break;
+			default:
+				number_of_objects = 1;
+				Warning(LOCATION, "There should only be NumberOfCraft of %s", object_type);	
```
This warning should say something like `NumberOfCraft for %s was %d but must be 1`.  Then use `number_of_objects` in the warning and set it after the warning.

Goober5000 commented on code/mission/import/xwingmissionparse.cpp:
```diff
@@ -623,6 +623,231 @@ void parse_xwi_flightgroup(mission *pm, const XWingMission *xwim, const XWMFligh
+			
+			/** Now convert the grid (a,b) to the relavenat formation ie. (x,y) or (z,y) etc **/
+			auto ojxyz = xwi_determine_mine_formation_position(oj, objectPosX, objectPosY, objectPosZ, offsetAxisA, offsetAxisB);
+			vm_vec_scale(&ojxyz, 1000); // units are in kilometers (after processing by xwinglib which handles the
```
Instead of doing the scaling here, do it when `objectPosX/Y/Z` are assigned.  Otherwise you'll have a mismatch between the grid distance in meters and the mine coordinates in kilometers.

Goober5000 commented on code/mission/import/xwingmissionparse.cpp:
```diff
@@ -623,6 +623,231 @@ void parse_xwi_flightgroup(mission *pm, const XWingMission *xwim, const XWMFligh
+										// factor of 160), so scale them up
+
+			p_object pobj;
+			SCP_totitle(pobj.name);
```
`SCP_totitle(pobj.name);` should be moved to the function that generates the name.

Goober5000 commented on code/mission/import/xwingmissionparse.cpp:
```diff
@@ -623,6 +623,231 @@ void parse_xwi_flightgroup(mission *pm, const XWingMission *xwim, const XWMFligh
+			
+			int team = 0; //default civilian/team none
+			if (number_of_objects > 1)
+				team = 2; //mines are always hostile
```
Don't use `0` and `2` as "magic number" indexes; you'll need to look up the team index from the name:
```
int team = Species_info[sip->species].default_iff;

if (the-object-is-a-mine)
{
	auto team_name = "Hostile";
	int index = iff_lookup(team_name);
	if (index >= 0)
		team = index;
	else
		Warning(LOCATION, "Could not find iff %s", team_name);
}
```

Goober5000 commented on code/mission/import/xwingmissionparse.cpp:
```diff
@@ -623,6 +623,231 @@ void parse_xwi_flightgroup(mission *pm, const XWingMission *xwim, const XWMFligh
+				Warning(LOCATION, "Unable to determine ship class for Object Group with type %s", object_type);
+				ship_class = 0;
+			}
+			auto sip = &Ship_info[ship_class];
```
Remove this from here since it's now at the top of the function

Goober5000 commented on code/mission/import/xwingmissionparse.cpp:
```diff
@@ -623,6 +623,231 @@ void parse_xwi_flightgroup(mission *pm, const XWingMission *xwim, const XWMFligh
+				if (n > objectNameSet.size()) { // if Size = all the elements + null then a name should be found
+					goto noNameError;
+				}
+			}
```
Most/all of this should be in the name picker function.  Call it `xwi_determine_space_object_name` rather than `unique_object_suffix`

Goober5000 commented on code/mission/import/xwingmissionparse.cpp:
```diff
@@ -623,6 +623,231 @@ void parse_xwi_flightgroup(mission *pm, const XWingMission *xwim, const XWMFligh
+			pobj.team = team;
+			pobj.initial_velocity = 0;
+
+			objectNameSet.insert(pobj.name); // add the new name to the objectSet
```
This line should be moved to where the name is actually assigned

Goober5000 commented on code/mission/import/xwingmissionparse.cpp:
```diff
@@ -623,6 +623,231 @@ void parse_xwi_flightgroup(mission *pm, const XWingMission *xwim, const XWMFligh
+	}
+noNameError:
+Warning(LOCATION, "Unable to find a suffix for %s", object_type);
+return;
```
This bit won't be needed

Dark-Visor committed:
Update to object naming function and cleanup

Move entirety of the object naming to it's own function
XWI_determine_space_object_name.

Cleanup the XWI_determine_object_type
to remove redundant case returns. Four cases for the mine have now been
grouped to have a single return.

Goober5000 commented on code/mission/import/xwingmissionparse.cpp:
```diff
@@ -623,6 +623,208 @@ void parse_xwi_flightgroup(mission *pm, const XWingMission *xwim, const XWMFligh
+	return;
+}
+
+// find lowest unique n for the suffix should return true when a unique name is found else false
```
This comment isn't accurate anymore (the function doesn't return true/false)

Goober5000 commented on code/mission/import/xwingmissionparse.cpp:
```diff
@@ -623,6 +623,208 @@ void parse_xwi_flightgroup(mission *pm, const XWingMission *xwim, const XWMFligh
+			case XWMObjectType::oj_Mine4:
+				break;
+			default:
+				Warning(LOCATION, "NumberOfCraft of %s", object_type, " was %d", number_of_objects, " but must be 1.");
```
The way Warning works, you need to have the string in one piece, and then all the variables in order after it

Goober5000 commented on code/mission/import/xwingmissionparse.cpp:
```diff
@@ -623,6 +623,208 @@ void parse_xwi_flightgroup(mission *pm, const XWingMission *xwim, const XWMFligh
+	// Copy objects in Parse_objects to set for name checking below
+	// This only needs to be done fully once per object group then can be added to after each new object
+	SCP_set<SCP_string> objectNameSet;
+	for (int n = 0; n < Parse_objects.size(); n++) {
```
This is going to cause a signed/unsigned warning, so you need to put an `(int)` cast in front of `Parse_objects.size()`

Goober5000 commented on code/mission/import/xwingmissionparse.cpp:
```diff
@@ -623,6 +623,208 @@ void parse_xwi_flightgroup(mission *pm, const XWingMission *xwim, const XWMFligh
+			
+			int team = Species_info[sip->species].default_iff;
+
+			if (number_of_objects > 1) {
```
Check that the object type is a mine here, instead of checking the number of objects.  You can do a switch on oj->objectType, same as in the beginning of the function.

Goober5000 commented on code/mission/import/xwingmissionparse.cpp:
```diff
@@ -623,6 +623,208 @@ void parse_xwi_flightgroup(mission *pm, const XWingMission *xwim, const XWMFligh
+			pobj.orient = orient;
+			pobj.pos = ojxyz;
+			
+			int team = Species_info[sip->species].default_iff;
```
Move this line and the following lines determining the team above the loops.  The team only needs to be determined once, since it is the same for all space objects in any group.

Goober5000 commented on code/mission/import/xwingmissionparse.cpp:
```diff
@@ -623,6 +623,208 @@ void parse_xwi_flightgroup(mission *pm, const XWingMission *xwim, const XWMFligh
+			pobj.team = team;
+			pobj.initial_velocity = 0;
+
+			Parse_objects.push_back(pobj);
```
The object will need its subsystems handled as well:
```
switch (oj->objectType) {
	case XWMObjectType::oj_Mine1:
	case XWMObjectType::oj_Mine2:
	case XWMObjectType::oj_Mine3:
	case XWMObjectType::oj_Mine4:
		pobj.subsys_index = Subsys_index;
		int this_subsys = allocate_subsys_status();
		pobj.subsys_count++;
		strcpy_s(Subsys_status[this_subsys].name, NOX("Pilot"));

		for (int n = 0; n < sip->n_subsystems; n++) {
			auto subsys = &sip->subsystems[n];
			if (subsys->type == SUBSYSTEM_TURRET) {
				this_subsys = allocate_subsys_status();
				pobj.subsys_count++;
				strcpy_s(Subsys_status[this_subsys].name, sip->subsystems[n].name);

				for (int bank = 0; bank < MAX_SHIP_PRIMARY_BANKS; bank++) {
					if (subsys->primary_banks[bank] >= 0) {
						Subsys_status[this_subsys].primary_banks[bank] = mine_laser_index;
					}
				}
			}
		}
		break;
	default:
		break;
}
```
And `mine_laser_index` will need to be determined near the top of the function - this will change the ion to the laser.

Dark-Visor committed:
Update mine weapon subsustem and other fixes

Update the mine weapon subsystem from ion to laser since the default in
FotG is ion as well as other code fixes and some cleanup.

Dark-Visor committed:
Correct switch statement for iff_lookup

The iff_lookup switch statement should include a break and default case.

Goober5000 commented on code/mission/import/xwingmissionparse.cpp:
```diff
@@ -623,6 +624,250 @@ void parse_xwi_flightgroup(mission *pm, const XWingMission *xwim, const XWMFligh
+	int mine_dist = 400; // change this to change the distance between the mines
+	/** The minefield is a square 2d along two planes (a,b) centred on the given position **/
+	float offsetAxisA = 0 -(mine_dist / 2 * (number_of_objects - 1)); // (- the distance to centre the grid)
+	float offsetAxisB = 0 -(mine_dist / 2 * (number_of_objects - 1));
```
If `number_of_objects` is 1, this will cause a div-0 error.  Make an if() block, and do the calculations only if `number_of_objects` > 1.  Otherwise, set the offsets to 0.

Goober5000 commented on code/mission/import/xwingmissionparse.cpp:
```diff
@@ -623,6 +624,250 @@ void parse_xwi_flightgroup(mission *pm, const XWingMission *xwim, const XWMFligh
+void parse_xwi_objectgroup(mission* pm, const XWingMission* xwim, const XWMObject* oj)
+{
+	SCP_UNUSED(pm);
+	auto object_type = xwi_determine_object_type(oj);
```
For clarity, and to be consistent with flight group parsing, rename `object_type` to `class_name` and rename `xwi_determine_object_type` to `xwi_determine_object_class`.

Goober5000 commented on code/mission/import/xwingmissionparse.cpp:
```diff
@@ -623,6 +624,250 @@ void parse_xwi_flightgroup(mission *pm, const XWingMission *xwim, const XWMFligh
+			auto ojxyz = xwi_determine_mine_formation_position(oj, objectPosX, objectPosY, objectPosZ, offsetAxisA, offsetAxisB);
+
+			// Defense Mine#Ion needs to have it's weapon changed from ion to laser
+			int mine_laser_index = weapon_info_lookup("T&B KX-5#imp");
```
This can be moved outside the two loops; the lookup only needs to happen once.

Also, it's -> its

Goober5000 commented on code/mission/import/xwingmissionparse.cpp:
```diff
@@ -623,6 +624,250 @@ void parse_xwi_flightgroup(mission *pm, const XWingMission *xwim, const XWMFligh
+	return iter.first->c_str();
+}
+
+void parse_xwi_objectgroup(mission* pm, const XWingMission* xwim, const XWMObject* oj)
```
Search the file for `// load flight groups` and add a loop for the object groups similar to the existing loop for flight groups

Goober5000 reviewed (changes requested):
Just a few more things

Dark-Visor committed:
Add objects to parse_xwi_mission and fixes

Add lines to parse_xwi_mission so that the objects would be included.
Also include some other fixes and code cleanup.

Goober5000 commented on code/mission/import/xwingmissionparse.cpp:
```diff
@@ -623,6 +624,238 @@ void parse_xwi_flightgroup(mission *pm, const XWingMission *xwim, const XWMFligh
+	float offsetAxisB = 0;
+
+	int mine_dist = 400; // change this to change the distance between the mines
+	int mine_laser_index = weapon_info_lookup("T&B KX-5#imp"); // "Defense Mine#Ion" needs to have its weapon changed to laser
```
If the index is < 0, add a warning that the weapon couldn't be found

Goober5000 commented on code/mission/import/xwingmissionparse.cpp:
```diff
@@ -623,6 +624,238 @@ void parse_xwi_flightgroup(mission *pm, const XWingMission *xwim, const XWMFligh
+			team = index;
+		else
+			Warning(LOCATION, "Could not find iff %s", team_name);
+		if (number_of_objects > 1) {
```
Put a blank line before this line.  The separation will help indicate that the two groups of lines are for different purposes

Goober5000 commented on code/mission/import/xwingmissionparse.cpp:
```diff
@@ -623,6 +624,238 @@ void parse_xwi_flightgroup(mission *pm, const XWingMission *xwim, const XWMFligh
+				pobj.subsys_count++;
+				strcpy_s(Subsys_status[this_subsys].name, NOX("Pilot"));
+
+				for (int n = 0; n < sip->n_subsystems; n++) {
```
Surround this loop with an if() that only performs these actions if `mine_laser_index` is >= 0.  This will avoid replacing the weapon if the laser can't be found

Dark-Visor committed:
Add warning for weapon change plus fixes

Ensure that the weapon change for mines is valid and only apply the
change to subsystems if valid. Add line spacing to improve clarity and
readability.

Goober5000 commented on code/mission/import/xwingmissionparse.cpp:
```diff
@@ -746,6 +746,9 @@ void parse_xwi_objectgroup(mission* pm, const XWingMission* xwim, const XWMObjec
 
 	int mine_dist = 400; // change this to change the distance between the mines
 	int mine_laser_index = weapon_info_lookup("T&B KX-5#imp"); // "Defense Mine#Ion" needs to have its weapon changed to laser
+	if (mine_laser_index < 1)
```
This should be <0, not <1.  The number 0 is a valid index.

Goober5000 commented on code/mission/import/xwingmissionparse.cpp:
```diff
@@ -826,19 +830,22 @@ void parse_xwi_objectgroup(mission* pm, const XWingMission* xwim, const XWMObjec
-						for (int bank = 0; bank < MAX_SHIP_PRIMARY_BANKS; bank++) {
-							if (subsys->primary_banks[bank] >= 0) {
-								Subsys_status[this_subsys].primary_banks[bank] = mine_laser_index;
+				if (mine_laser_index > 0) {
```
This should be >= 0, not >.  Again, 0 itself is a valid index

Goober5000 commented on code/mission/import/xwingmissionparse.cpp:
```diff
@@ -623,6 +624,245 @@ void parse_xwi_flightgroup(mission *pm, const XWingMission *xwim, const XWMFligh
+	float offsetAxisB = 0;
+
+	int mine_dist = 400; // change this to change the distance between the mines
+	int mine_laser_index = weapon_info_lookup("T&B KX-5#imp"); // "Defense Mine#Ion" needs to have its weapon changed to laser
```
As long as you're here, add this:
```
auto weapon_name = "T&B KX-5#imp";
```
Then use `weapon_name` in both the lookup and the warning.  See how `team_name` is handled below.

Dark-Visor committed:
Update mine_laser_index

mine_laser_index should be treated similar to team_name index and
ship_class with a warning to notify if the index int is not valid/less
than 0. Add auto weapon_name so that the warning can display the name of
the weapon in the output.

Dark-Visor committed:
Fix switch cases to be in a block

Encase the declared variables in the switch cases inside blocks to avoid
compile errors.
